providers/implementations/signature/dsa_sig.c

```diff
@@ -562,6 +578,190 @@ static const OSSL_PARAM *dsa_settable_ctx_md_params(void *vpdsactx)
+        if (bld == NULL)
+            goto err;
+
+        bnctx = BN_CTX_new_ex(libctx);
+        if (bnctx == NULL)
+            goto err;
+
+        for (p = t->key; p->data != NULL; p++) {
+            switch (p->type) {
+            case OSSL_PARAM_UNSIGNED_INTEGER:
+                {
+                    BIGNUM *bn = BN_CTX_get(bnctx);
+                    if (bn == NULL
+                        || BN_bin2bn(p->data, p->data_len, bn) == NULL
+                        || !OSSL_PARAM_BLD_push_BN(bld, p->name, bn))
+                        goto err;
+                    break;
+                }
+            case OSSL_PARAM_OCTET_STRING:
+                if (!OSSL_PARAM_BLD_push_octet_string(bld, p->name, p->data,
+                                                      p->data_len))
+                    goto err;
+                break;
+            case OSSL_PARAM_UTF8_STRING:
+                if (!OSSL_PARAM_BLD_push_utf8_string(bld, p->name, p->data,
+                                                     p->data_len))
+                    goto err;
+                break;
+            }
+        }
+
+        key = OSSL_PARAM_BLD_to_param(bld);
+        if (key == NULL)
+            goto err;
+
+        OSSL_PARAM_BLD_free(bld);
+        bld = NULL;
+
+        BN_CTX_free(bnctx);
+        bnctx = NULL;
+
+        ctx = dsa_newctx(&test_provider, NULL);
+        if (ctx == NULL)
+            goto err;
+
+        provkey = dsa_newdata(&test_provider);
+        if (provkey == NULL)
+            goto err;
+
+        if (!dsa_import(provkey, OSSL_KEYMGMT_SELECT_KEYPAIR
+                        | OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS, key))
+            goto err;
+
+        OSSL_PARAM_free(key);
+        key = NULL;
+
+        if (!dsa_digest_sign_init(ctx, t->mdalgorithm, provkey, NULL))
+            goto err;
+        if (!dsa_digest_signverify_update(ctx, (const unsigned char *)msg,
+                                          strlen(msg)))
+            goto err;
+        siglen = sizeof(sig) / sizeof(*sig);
+        if (!dsa_digest_sign_final(ctx, sig, &siglen, siglen))
+            goto err;
+        if (!dsa_digest_verify_init(ctx, t->mdalgorithm, provkey, NULL))
+            goto err;
+        if (!dsa_digest_signverify_update(ctx, (const unsigned char *)msg,
+                                          strlen(msg)))
+            goto err;
+        if (t->sig_expected != NULL
+            && (siglen != t->sig_expected_len
+                || memcmp(sig, t->sig_expected, t->sig_expected_len) != 0))
+            goto err;
+        if (!dsa_digest_verify_final(ctx, sig, siglen))
+            goto err;
+
+        dsa_freedata(provkey);
+        provkey = NULL;
+        dsa_freectx(ctx);
+        ctx = NULL;
+    }
+
+    success = 1;
+err:
+    /* cache result for next invocation */
+    memset(&selftest_thread, 0, sizeof(selftest_thread));
+    selftest_status = success;
+    CRYPTO_THREAD_unlock(selftest_lock);
+
+    if (bnctx)
```

Comment:
Our coding style requires the `!= NULL` when testing pointers.  However, in this case it's not necessary.  All of our free functions should correctly handle a NULL pointer being passed so lose the if statements.